File: src/TelaInicial.hs
module TelaInicial where


import Usuario (Usuario(..), modificarUsuario)
import Tarefa (Tarefa, listarTarefasUsuario, backlogEmpresa)
import Sprint (Sprint, listarSprint, listarSprintsDaEmpresa)

-- Menu para Administrador
menuAdministrador :: Usuario -> [Usuario]-> [Tarefa] -> [Sprint] -> IO ()
menuAdministrador usuario usuarios tarefas sprints = do
  putStrLn "\nMenu Administrador:"
  putStrLn "1. Backlog"
  putStrLn "2. Sprint"
  putStrLn "3. Dados do Usuario"
  putStrLn "4. Logout"
  escolha <- getLine
  case escolha of
    "1" -> do
      tarefasAtualizado <- backlogEmpresa usuario tarefas
      menuAdministrador usuario usuarios tarefasAtualizado sprints
    "2" -> do
      (sprintsAtualizada,tarefasAtualizado) <- listarSprintsDaEmpresa usuario sprints usuarios tarefas
      menuAdministrador usuario  usuarios tarefas sprints
    "3" -> do
      usuarioAtualizado <- modificarUsuario usuario
      let usuariosAtualizados = map (\u -> if usuarioId u == usuarioId usuario then usuarioAtualizado else u) usuarios
      menuAdministrador usuarioAtualizado usuarios tarefas sprints
    "4" -> putStrLn "Saindo..."
    _   -> do
      putStrLn "Opção inválida, tente novamente."
      menuAdministrador usuario usuarios tarefas sprints

-- Menu para Usuário Comum
menuComum :: Usuario ->[Usuario] -> [Tarefa] -> [Sprint] -> IO ()
menuComum usuario usuarios tarefas sprints = do
  putStrLn "\nMenu Usuário Comum:"
  putStrLn "1. Sprints"
  putStrLn "2. Tarefas"
  putStrLn "3. Dados do Usuario"
  putStrLn "4. Logout"
  escolha <- getLine
  case escolha of
    "1" -> do
      listarSprint usuario sprints tarefas
      menuComum usuario usuarios tarefas sprints
    "2" -> do
      listarTarefasUsuario usuario tarefas
      menuComum usuario usuarios tarefas sprints
    "3" -> do
      usuarioAtualizado <- modificarUsuario usuario
      let usuariosAtualizados = map (\u -> if usuarioId u == usuarioId usuario then usuarioAtualizado else u) usuarios
      menuComum usuarioAtualizado usuariosAtualizados tarefas sprints
    "4" -> putStrLn "Saindo..."
    _   -> do
      putStrLn "Opção inválida, tente novamente."
      menuComum usuario usuarios tarefas sprints

File: prolog/src/tela_inicial.pl
:- module(tela_inicial, [menu_administrador/1, menu_comum/1]).

:- use_module(sprint).
:- use_module(tarefa).
:- use_module(usuario).

% Menu para usuários administradores
menu_administrador(Usuario) :-
    writeln('\nMenu Administrador:'),
    writeln('1. Backlog'),
    writeln('2. Sprint'),
    writeln('3. Dados do Usuário'),
    writeln('4. Logout'),
    read_line_to_string(user_input, EscolhaString),
    atom_number(EscolhaString, Escolha),  % Converte a string para número
    (   Escolha = 1 -> 
        listar_tarefas_backlog(Usuario),
        menu_administrador(Usuario)
    ;   Escolha = 2 -> 
        listar_sprints_da_empresa(Usuario),
        menu_administrador(Usuario)
    ;   Escolha = 3 -> 
        modificar_usuario(Usuario, UsuarioAtualizado),
        menu_administrador(UsuarioAtualizado)
    ;   Escolha = 4 -> 
        null
    ;   
        writeln('Opção inválida, tente novamente.'),
        menu_administrador(Usuario)
    ).

% Menu para usuários comuns
menu_comum(Usuario) :-
    writeln('\nMenu Usuário Comum:'),
    writeln('1. Sprints'),
    writeln('2. Tarefas'),
    writeln('3. Dados do Usuário'),
    writeln('4. Logout'),
    read_line_to_string(user_input, EscolhaString),
    atom_number(EscolhaString, Escolha),  % Converte a string para número
    (   Escolha = 1 -> 
        listar_sprints_da_empresa(Usuario),
        menu_comum(Usuario)
    ;   Escolha = 2 -> 
        listar_tarefas_usuario(Usuario),
        menu_comum(Usuario)
    ;   Escolha = 3 -> 
        modificar_usuario(Usuario, UsuarioAtualizado),
        menu_comum(UsuarioAtualizado)
    ;   Escolha = 4 -> 
        writeln('Saindo do menu...'),  % Mensagem opcional
        true  % Sai do menu sem repetir
    ;   
        writeln('Opção inválida, tente novamente.'),
        menu_comum(Usuario)  % Repete o menu em caso de erro
    ).

% Atualiza os dados do usuário
update_usuario(Usuario, UsuarioAtualizado) :-
    findall(U, usuario(U), Usuarios),  % Obtém a lista de todos os usuários
    maplist(replace_usuario(Usuario, UsuarioAtualizado), Usuarios, UsuariosAtualizados),
    retractall(usuario(_)),  % Remove todos os usuários antigos
    maplist(assertz, UsuariosAtualizados).  % Adiciona a lista atualizada de usuários

% Predicado auxiliar para substituir um usuário
replace_usuario(Usuario, UsuarioAtualizado, Usuario, UsuarioAtualizado) :- !.  % Substitui o usuário
replace_usuario(_, _, U, U).  % Mantém os outros usuários
